SmartDeviceLink/SDLUploadFileOperation.m

```diff
@@ -51,112 +51,187 @@ - (instancetype)initWithFile:(SDLFileWrapper *)file connectionManager:(id<SDLCon
 
 - (void)start {
     [super start];
-
-    [self sdl_sendPutFiles:[self.class sdl_splitFile:self.fileWrapper.file mtuSize:[SDLGlobals sharedGlobals].maxMTUSize] withCompletion:self.fileWrapper.completionHandler];
+    [self sendPutFiles:self.fileWrapper.file mtuSize:[SDLGlobals sharedGlobals].maxMTUSize withCompletion:self.fileWrapper.completionHandler];
 }
 
-- (void)sdl_sendPutFiles:(NSArray<SDLPutFile *> *)putFiles withCompletion:(SDLFileManagerUploadCompletionHandler)completion {
+/**
+ Sends data asychronously to the SDL Core by breaking the data into smaller packets, each of which is sent via a putfile. To prevent large files from eating up memory, the data packet is deleted once it is sent via a putfile. If the SDL Core receives all the putfiles successfully, a success response with the amount of free storage space left on the SDL Core is returned. Otherwise the error returned by the SDL Core is passed along.
+
+ @param file The file containing the data to be sent to the SDL Core
+ @param mtuSize The maximum packet size allowed
+ @param completion Closure returning whether or not the upload was a success
+ */
+- (void)sendPutFiles:(SDLFile *)file mtuSize:(NSUInteger)mtuSize withCompletion:(SDLFileManagerUploadCompletionHandler)completion  {
+    dispatch_group_t putFileGroup = dispatch_group_create();
+    dispatch_group_enter(putFileGroup);
+    __weak typeof(self) weakself = self;
     __block BOOL stop = NO;
     __block NSError *streamError = nil;
     __block NSUInteger bytesAvailable = 0;
     __block NSInteger highestCorrelationIDReceived = -1;
 
-    dispatch_group_t putFileGroup = dispatch_group_create();
-    dispatch_group_enter(putFileGroup);
-
-    // When the putfiles all complete, run this block
-    __weak typeof(self) weakself = self;
+    // Waits for all packets be sent before returning whether or not the upload was a success
     dispatch_group_notify(putFileGroup, dispatch_get_main_queue(), ^{
         if (streamError != nil || stop) {
             completion(NO, bytesAvailable, streamError);
         } else {
             completion(YES, bytesAvailable, nil);
         }
-
         [weakself finishOperation];
     });
 
-    for (SDLPutFile *putFile in putFiles) {
+    [self openInputStreamWithFile:file];
+
+    // Break the data into small pieces, each of which will be sent in a separate putfile
+    unsigned long long fileSize = [file fileSize];
+    NSUInteger currentOffset = 0;
+    unsigned long long numberOfFilesToSend = (((fileSize - 1) / mtuSize) + 1);
+    for (int i = 0; i < numberOfFilesToSend; i++) {
```

Comment:
I consolidated some of the code. I also broke out a lot of logic into separate functions.